src/org/zaproxy/zap/extension/proxies/ExtensionProxies.java

```diff
@@ -0,0 +1,226 @@
+
+    @Override
+    public void optionsLoaded() {
+        restartProxies();
+    }
+
+    @Override
+    public void optionsChanged(OptionsParam optionsParam) {
+        restartProxies();
+    }
+
+    private void restartProxies() {
+        List<ProxiesParamProxy> proxyParams = this.getParam().getProxies();
+        Map<String, ProxyServer> currentProxies = proxyServers;
+        proxyServers = new HashMap<String, ProxyServer>();
+        for (ProxiesParamProxy proxyParam : proxyParams) {
+            if (proxyParam.isEnabled()) {
+                // Treat disabled proxies as if they dont really exist
+                String key = proxyParam.getAddress() + ":" + proxyParam.getPort();
+                ProxyServer proxy = currentProxies.remove(key);
+                if (proxy == null) {
+                    // Its a new one
+                    proxy = startProxyServer(proxyParam.getAddress(), proxyParam.getPort());
+                }
+                proxyServers.put(key, proxy);
+            }
+        }
+        // Any proxies left have been removed
+        for (Entry<String, ProxyServer> entry : currentProxies.entrySet()) {
+            stopProxyServer(entry.getValue());
+        }
+    }
+
+    private ProxyServer startProxyServer(String address, int port) {
+        // Note that id this is _not_ set then the proxy will go into a nasty loop if you point a browser at it
+        String key = address + ":" + port;
+        log.info("Starting alt proxy server: " + key);
+        ProxyServer proxyServer = new ProxyServer(ZAP_PROXY_THREAD_PREFIX + key);
+        proxyServer.setEnableApi(true);
+        if (proxyServer.startServer(address, port, false) > 0) {
+            Control.getSingleton().getExtensionLoader().getExtension(ExtensionHistory.class).registerProxy(proxyServer);
+        }
+        return proxyServer;
+    }
+
+    private void stopProxyServer(ProxyServer proxyServer) {
+        log.info("Stopping alt proxy server: " + proxyServer.getProxyParam().getProxyIp());
+        Control.getSingleton().getExtensionLoader().getExtension(ExtensionHistory.class).unregisterProxy(proxyServer);
+        proxyServer.stopServer();
+    }
+
+    public List<ProxiesParamProxy> getAdditionalProxies() {
+        return this.getParam().getProxies();
+    }
+
+    public ProxiesParamProxy getAdditionalProxy(String address, int port) {
+        for (ProxiesParamProxy p : this.getParam().getProxies()) {
+            if (p.getAddress().equals(address) && p.getPort() == port) {
+                return new ProxiesParamProxy(p);
+            }
+        }
+        return null;
+    }
+
+    public void addProxy(ProxiesParamProxy proxy) {
+        if (this.getAdditionalProxy(proxy.getAddress(), proxy.getPort()) != null) {
+            throw new IllegalArgumentException("Proxy already exists");
+        }
+        if (!this.canListenOn(proxy.getAddress(), proxy.getPort())) {
+            throw new IllegalArgumentException("Cannot listen on address:port");
```

Comment:
Might be worth including the attempted address:port

Reply:
yeah